File: 450-DSA-SHEET/ARRAYS/Python/4-Sort012.py
def sort012(arr, n):
    low = 0
    mid = 0
    high = n-1

    while mid <= high:
        if arr[mid] == 0:
            arr[low], arr[mid] = arr[mid], arr[low]
            low = low+1
            mid = mid+1

        elif arr[mid] == 1:
            mid = mid+1

        else:
            arr[mid], arr[high] = arr[high], arr[mid]
            high = high-1
    
    return arr


def printArray(arr):
    for i in range(0,n):
        print(arr, end = ' ')


arr = [0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1]
n = len(arr)
sort012(arr, n)
printArray(arr)

File: 450-DSA-SHEET/ARRAYS/C-Plus-Plus/4-Sort012.cpp
#include <bits/stdc++.h>
using namespace std;


void sort012(int arr[], int n){
    int low = 0, mid = 0;
    int high = n-1;
    while(mid<=high){
        switch(arr[mid]){
            case 0: 
                swap(arr[low++], arr[mid++]);
                break;

            case 1:
                mid++;
                break;
            
            case 2:
                swap(arr[mid], arr[high--]);
                break;
        }
    }
}

int main(){
    int n; 
    cout << "Enter the number of elements: ";
    cin>>n;
    int arr[n];
    cout << "Enter the elements of the array: ";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    sort012(arr,n);
    cout << "DUTCH NATIONAL FLAG PROBLEM: " << endl;
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout << endl;
    // 0, 1, 1, 0, 1, 2, 1, 2, 0, 0, 0, 1
    return 0;
}